### ISSUE TITLE
Add an alternative to (or modify the behavior of) n in json_pack()

### ISSUE THREAD

**Issue by HalosGhost** [akheron/jansson#261]
This is a pain point I have run into with jansson (I find most of the rest of the library to be really fantastic quality): the ``n`` specifier does not expect/consume an argument. I can come up with a bunch of reasons why this makes sense, and I am sure it is radically easier to create. However, this leads to times where adding a nullable value can become quite a pain. Namely, I have felt forced into creating large conditional blocks to correctly assign them ``null`` or another value:

```c
if ( label_j && s->secs ) {
    new_paste = json_pack("{s:s,s:s,s:s,s:s}", "long", lid, "msg", msg,
                          "label", label, "sunset", sunset);
} else if ( label_j && !s->secs ) {
    new_paste = json_pack("{s:s,s:s,s:s,s:n}", "long", lid, "msg", msg,
                          "label", label, "sunset");
} else if ( !label_j && s->secs ) {
    new_paste = json_pack("{s:s,s:s,s:n,s:s}", "long", lid, "msg", msg,
                          "label", "sunset", sunset);
} else {
    new_paste = json_pack("{s:s,s:s,s:n,s:n}", "long", lid, "msg", msg,
                          "label", "sunset");
}
```

If there were, for example, a version of the ``n`` specifier that expected/consumed an argument but ignored it, I could, at least, create the specifier list in a conditional and then make only one call to ``json_pack()``:

```c
const char * spec =  label_j &&  s->secs ? "{s:s,s:s,s:s,s:s}" :
                     label_j && !s->secs ? "{s:s,s:s,s:s,s:n}" :
                    !label_j &&  s->secs ? "{s:s,s:s,s:n,s:s}" :
                                           "{s:s,s:s,s:n,s:n}" ;
new_paste = json_pack(spec, "long", lid, "msg", msg, "label", label, "sunset", sunset);
```

Another alternative (perhaps even cleaner) would be to add something like the ``?`` modifier to values to explicitly mark them as nullable, allowing the caller to pass ``NULL`` in place of a variable of the otherwise-expected type. Doing so would actually allow for the removal of a conditional entirely:

```c
new_paste = json_pack("{s:s,s:s,s:s?,s:s?}", "long", lid, "msg", msg, "label", label, "sunset", sunset);
```
(assuming that ``label`` and ``sunset`` are set to ``NULL`` as they need to be).

Of course, this adds some complexity for non-string types because ``NULL`` is mostly just ``0`` for other types. Perhaps a first step toward doing this would be to add something like an ``S`` specifier for strings that could be ``null``?

---
It is also possible that I have misread the API reference and there is some cleaner way of accomplishing what I want already; if so, please point me to where in the docs it is covered. But, as far as I can tell, this is one of the few cases where jansson is not quite as flexible as I would like.

**Comment by akheron** [comment]
I like the `?` specifier to have nullable fields. `?` would only be valid after `o`, `O` and `s` because they accept pointers. Technically, `s#`, `s%`, `+`, `+#` and `+%` also accept pointers, but how they are used probably leads to the argument always being non-NULL.

**Comment by HalosGhost** [comment]
in my personal opinion, the ``?`` specifier is the best solution here (though admittedly something like ``N``, which would just be ``n`` but it consumes an argument while ignoring it, would also be a great step forward).

I am really glad to hear you are open to this! The initial snippet above is actually taken from one of my personal projects, the second two snippets are the pseudo-code that I would want to write.

**Comment by akheron** [comment]
Implemented in #270
